stdlib/buffer.mli

```diff
@@ -30,6 +30,18 @@
 
 *)
 
+(** {b Concurrency safety} *)
+
+[@@@alert concurrency_unsafe
+    "Concurrent use of buffers is a programming error."
```

Comment:
Thanks for the PR to document the concurrency safety. Putting myself into the shoes of a user, I feel that the current documentation errs too much on the side of safety and makes the module unusable for reasonable use cases. 

We are conflating concurrency and parallelism when we mean `concurrency_unsafe`. To be precise, we have three modes of concurrency in OCaml 5: fibers, systhreads and domains. OCaml 4 had 2: systhreads and monadic concurrency libraries such as Lwt and Async. 

The observation is that if the user only uses fibers, with a library such as `eio`, then the Buffer module is safe. In particular, anything that is safe with Lwt (modulo the parts that uses systhreads -- `Lwt_preemptive` and others?) should remain safe with fibers.

Reply:
I could rename the section and alert to `Parallelism safety` or `Domain safety`? Thread safety might be confusing with `systhreads`  being a concurrency mechanism.

Similarly, if we want to de-emphasize those warnings, I can remove the alert and move the section to the end of the module documentations (with an example of a potential issue?). That sounds sensible to me since I am not sure if people will ever enable the alert since it is such an imprecise tool. Note however, that the alert is disabled by default, thus the modules are still perfectly usable.

Reply:
It is possible to add alerts to individual functions (or documentation comments) when needed.
For arrays, do you meant the `blit` function?

In term of users documentation, iI was draw to a hierarchy of the form:

1. no crashes
2. no broken abstractions
3. linearizability

since corrupting a data structure (breaking abstractions and expected invariants) is a surprising result for a sequential OCaml programmer.
 
If I am not mistaken, the first level of safety ought to be guarantee by the language, so it is better documented in the future memory model section of the manual to focus on module specific in the module-by-module documentation.

Reply:
> For arrays, do you meant the blit function?

I did have the `Array.blit` function in mind. In particular, `Array.blit` on float arrays does not respect the memory model. 

https://github.com/ocaml/ocaml/blob/66128baee81e1fd43ea1fbddbbebd0c3d0acc20e/runtime/array.c#L345-L356